amethyst_gltf/src/systems.rs

```diff
@@ -0,0 +1,410 @@
+            texture_storage,
+            material_defaults,
+            mesh_handles,
+            texture_handles,
+        );
+    }
+
+    if node.primitives.len() > 0 {
+        // If we only have a single graphics primitive, we load it onto the nodes entity
+        if node.primitives.len() == 1 {
+            load_primitive(
+                node_index,
+                0,
+                scene_asset,
+                &node.primitives[0],
+                node_entity,
+                loader,
+                meshes,
+                mesh_storage,
+                materials,
+                texture_storage,
+                material_defaults,
+                mesh_handles,
+                texture_handles,
+            );
+        } else {
+            // If we have multiple graphics primitives, we need to add child entities for each
+            // primitive and load the graphics onto those
+            for (primitive_index, primitive) in node.primitives.iter().enumerate() {
+                let primitive_entity = entities.create();
+                local_transforms.insert(primitive_entity, LocalTransform::default());
+                transforms.insert(primitive_entity, Transform::default());
+                parents.insert(primitive_entity, Child::new(*node_entity));
+                load_primitive(
+                    node_index,
+                    primitive_index,
+                    scene_asset,
+                    primitive,
+                    &primitive_entity,
+                    loader,
+                    meshes,
+                    mesh_storage,
+                    materials,
+                    texture_storage,
+                    material_defaults,
+                    mesh_handles,
+                    texture_handles,
+                );
+            }
+        }
+    }
+}
+
+// Load a single graphics primitive, attach all data to the given `entity`.
+fn load_primitive(
+    node_index: usize,
+    primitive_index: usize,
+    scene_asset: &GltfSceneAsset,
+    primitive: &GltfPrimitive,
+    entity: &Entity,
+    loader: &Loader,
+    meshes: &mut WriteStorage<Handle<Mesh>>,
+    mesh_storage: &AssetStorage<Mesh>,
+    materials: &mut WriteStorage<Material>,
+    texture_storage: &AssetStorage<Texture>,
+    material_defaults: &MaterialDefaults,
+    mesh_handles: &mut Vec<(usize, usize, Handle<Mesh>)>,
+    texture_handles: &mut Vec<(usize, usize, Handle<Texture>)>,
+) {
+    let mesh = match primitive.handle {
+        // If we have a cached mesh handle, use that
+        Some(ref handle) => handle.clone(),
+
+        // Otherwise, load the mesh and request caching of the mesh handle
+        None => {
+            let handle = loader.load_from_data(primitive.attributes.clone().into(), mesh_storage);
+            mesh_handles.push((node_index, primitive_index, handle.clone()));
+            handle
+        }
+    };
+
+    // Load material for the primitive
+    let material = match primitive.material {
```

Comment:
I think it would be nicer to do this:

```rust
let material = primitive
    .material
    .map(|i| (i, scene_asset.materials.get(i)))
    .and_then(|(i, m)| load_material(i, m, ..))
    .unwrap_or_else(|| mat_defaults.0.clone());
```

Reply:
Hmm, i need to do unwrap on m in and_then then

Reply:
No?

Reply:
If I reverse it it works though.